### PR TITLE
[FABCJ-214] - External Java chaincode

### PR DESCRIPTION
**Pull request by OYamkovyy**
Actually, this MR is pretty the same as https://github.com/hyperledger/fabric-chaincode-java/pull/63 (was forked from there) but with all the updates from upstream and with resolved conflicts

### CODE REVIEW

File: examples/fabric-chaincode-example-external/build.gradle [hunk 1]
```diff
@@ -0,0 +1,246 @@
+/*
```
Comment (mbwhite) [inline examples/fabric-chaincode-example-external/build.gradle:1]: Great to have an example!  

Think it's worth sliming down the gralde file - it's got a lot of things in it copied from the main build's gradle file. Not needed.

**Comment by mbwhite** [comment]
Sorry - finally getting to review this...

**Review by mbwhite — CHANGES_REQUESTED** [review]
The application class; in nodejs this is within the chaincode library itself.

Could it be in this case aswell?

**Comment by mbwhite** [comment]
Just making notes as I go... I'm trying this out - and I've notice the environment variables used by the Java are different from the Node... (subtly) wondering if we could make the 2 consistent

File: fabric-chaincode-shim/src/main/java/org/hyperledger/fabric/shim/ChatChaincodeWithPeer.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    private ChaincodeBase chaincodeBase;
+
+    ChatChaincodeWithPeer(final ChaincodeBase chaincodeBase) throws IOException {
+        if (chaincodeBase == null) {
+            throw new IOException("chaincodeBase can't be null");
+        }
+        chaincodeBase.validateOptions();
+
+        this.chaincodeBase = chaincodeBase;
```
Comment (mbwhite) [inline fabric-chaincode-shim/src/main/java/org/hyperledger/fabric/shim/ChatChaincodeWithPeer.java:27]: In the ContractRouter.main() method there are some other calls that need to made on the ContractRouter before it's able to work.  This looks like the correct place to put the,.

File: examples/fabric-chaincode-example-external/src/main/java/org/hyperledger/fabric/example/Application.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        ChaincodeServerProperties chaincodeServerProperties = new ChaincodeServerProperties();
+
+        final String portChaincodeServer = System.getenv(PORT_CHAINCODE_SERVER);
+        if (portChaincodeServer == null || portChaincodeServer.isEmpty()) {
+            throw new IOException("chaincode server port not defined in system env. for example 'PORT_CHAINCODE_SERVER=9999'");
+        }
+        final int port = Integer.parseInt(portChaincodeServer);
+        chaincodeServerProperties.setPortChaincodeServer(port);
+
+        final String corePeerAddress = System.getenv(CORE_PEER_ADDRESS);
```
Comment (mbwhite) [inline examples/fabric-chaincode-example-external/src/main/java/org/hyperledger/fabric/example/Application.java:45]: Don't believe this really should be required for this?

**Review by mbwhite — CHANGES_REQUESTED** [review]
Afraid that I was unable to get this to work simple contract.
Can you take a look please.

File: fabric-chaincode-shim/src/main/java/org/hyperledger/fabric/shim/ChatChaincodeWithPeer.java (hunk 2 above)
Comment (OYamkovyy) [reply fabric-chaincode-shim/src/main/java/org/hyperledger/fabric/shim/ChatChaincodeWithPeer.java:27]: I agree that it's a good place
but I see also two issues there - those methods called in ContractRouter.main() are private and protected - not available there and I'm not sure if extending access modifiers for them is a good idea
2) here we work with ChaincodeBase abstract class and if in the future there will be another implementation of it (not only ContractRouter) - there're possible troubles IMO

what if just create one more public method in ContractRouter which will contain calls from .main() and will allow chaincode developers to start routing manually in chaincode together with starting server?

WDYT, @mbwhite?

Comment (mbwhite) [reply fabric-chaincode-shim/src/main/java/org/hyperledger/fabric/shim/ChatChaincodeWithPeer.java:27]: Yes; I agree with the solution you've suggested - what I would do myself. 

There won't be another subclass; the Contract Router is the preferred way forward.

**Comment by mbwhite** [comment]
@OYamkovyy is this ready to re-review? Thanks

**Comment by OYamkovyy** [comment]
> @OYamkovyy is this ready to re-review? Thanks

Hi, @mbwhite 
yes, I updated an example, it's working, we tested it on our env
and added a public method to ContractRouter as I proposed before
as well you asked to make env variables in the example consistent with nodejs, I renamed chaincode_id, but left port as is - there should be int value, and CORE_CHAINCODE_ADDRESS can be confusing
so, now it's ready to review again

**Comment by mbwhite** [comment]
@OYamkovyy thanks for the updated; sorry for the delay in reviewing. I'm just seeing if I can run this now.. not managed to get it work just yet, but not sure if that's a mistake in my configuration or not. 

I did notice though that if the error message if CORE_CHAINCODE_ID is missing doesn't match.

**Comment by OYamkovyy** [comment]
updated an error message

I don't know if it will be useful, there is a dockerized example from repo (oyamkovyi/java-chaincode-test:0.17) (the only difference is that variables named CHAINCODE_ADDRESS and CHAINCODE_ID). We used this image for testing and it works

**Comment by mbwhite** [comment]
thanks for the new commits.. I've got a bit further in testing this... 

My when I attempt to send a transaction to the Java CC server, I get output like this
 ```
16:23:29:642 INFO    org.hyperledger.fabric.shim.ChaincodeBase newChannelBuilder                      Configuring channel connection to peer.
Thread[grpc-default-executor-0,5,main] 16:23:29:642 INFO    org.hyperledger.fabric.shim.ChaincodeBase getMaxInboundMessageSize               Maximum Inbound Message Size [MAX_INBOUND_MESSAGE_SIZE] = 104857600
Thread[grpc-default-executor-0,5,main] 16:23:29:701 INFO    org.hyperledger.fabric.shim.impl.InvocationTaskManager <init>                    Max Pool Size [TP_MAX_POOL_SIZE]5
Thread[grpc-default-executor-0,5,main] 16:23:29:703 INFO    org.hyperledger.fabric.shim.impl.InvocationTaskManager <init>                    Queue Size [TP_CORE_POOL_SIZE]5000
Thread[grpc-default-executor-0,5,main] 16:23:29:704 INFO    org.hyperledger.fabric.shim.impl.InvocationTaskManager <init>                    Core Pool Size [TP_QUEUE_SIZE]5
Thread[grpc-default-executor-0,5,main] 16:23:29:705 INFO    org.hyperledger.fabric.shim.impl.InvocationTaskManager <init>                    Keep Alive Time [TP_KEEP_ALIVE_MS]5000
Thread[grpc-default-executor-0,5,main] 16:23:29:707 INFO    org.hyperledger.fabric.shim.impl.InvocationTaskExecutor <init>                   Thread pool created
Thread[grpc-default-executor-0,5,main] 16:23:29:820 INFO    org.hyperledger.fabric.shim.impl.InvocationTaskManager register                  Registering new chaincode name: "javacc:30e755c332d780fb6d5c55e9c48e5865fe22df2e642ecc228ef37c0d523793f9"

```

But I can't send any transactions... will try and dig into what's going on.

**Comment by mbwhite** [comment]
@OYamkovyy   regarding my previous comment.... found the problem with the docker container I was using... not the code

If you can get a clean build; then I'll merge this :-)

**Comment by OYamkovyy** [comment]
thanks for the feedback, @mbwhite 
I'm trying to fix a pipeline, but unsuccessfully for now
according to the report, there're 4 failing tests, but they're passing on my local env - I can't reproduce this failure in order to fix it
and another trouble - not enough test coverage on classes RoutingRegistryImpl.java and SerializerRegistryImpl.java which even was not affected by the PR. Looks like some issue with configuration - investigating it as well

so, if there're some ideas on how to deal with the issues above - let me know pls

**Comment by mbwhite** [comment]
Shall investigate @OYamkovyy  :-)

**Comment by OYamkovyy** [comment]
hi, @mbwhite 
sorry for spamming commits, but finally, I got this pipeline passed :-)
in some tests, env variables were being set, and not cleared out, that's why some other test failed
It a strange thing that in other environments there was not such a problem, but now it works and that's cool 👍

**Review by mbwhite — APPROVED** [review]
Thankyou

**Pull request merged**